streams/conversion.ts

```diff
@@ -196,6 +196,78 @@ export function readableStreamFromIterable<T>(
   });
 }
 
+/**
+ * Convert the generator function into a TransformStream.
+ *
+ * ```ts
+ * import { readableStreamFromIterable, transformStreamFromGeneratorFunction } from "https://deno.land/std@$STD_VERSION/streams/mod.ts";
+ *
+ * const readable = readableStreamFromIterable([0, 1, 2])
+ *   .pipeThrough(transformStreamFromGeneratorFunction(async function* (src) {
+ *     for await (const chunk of src) {
+ *       yield chunk * 100;
+ *     }
+ *   }));
+ *
+ * for await (const chunk of readable) {
+ *   console.log(chunk);
+ * }
+ * // output: 0, 100, 200
+ * ```
+ *
+ * @param transformer A function to transform.
+ * @param writableStrategy An object that optionally defines a queuing strategy for the stream.
+ * @param readableStrategy An object that optionally defines a queuing strategy for the stream.
+ */
+export function transformStreamFromGeneratorFunction<I, O>(
+  transformer: (src: ReadableStream<I>) => Iterable<O> | AsyncIterable<O>,
+  writableStrategy?: QueuingStrategy<I>,
+  readableStrategy?: QueuingStrategy<O>,
+): TransformStream<I, O> {
+  const {
+    writable,
+    readable,
+  } = new TransformStream<I, I>(undefined, writableStrategy);
+
+  const iterable = transformer(readable);
+  const iterator: Iterator<O> | AsyncIterator<O> =
+    (iterable as AsyncIterable<O>)[Symbol.asyncIterator]?.() ??
+      (iterable as Iterable<O>)[Symbol.iterator]?.();
+  return {
+    writable,
+    readable: new ReadableStream<O>({
+      async pull(controller) {
+        let result: IteratorResult<O>;
+        try {
+          result = await iterator.next();
+        } catch (error) {
+          // Propagate error to stream from iterator
+          // If the stream status is "errored", it will be thrown, but ignore.
+          await readable.cancel(error).catch();
+          controller.error(error);
+          return;
+        }
+        if (result.done) {
+          controller.close();
+          return;
+        }
+        controller.enqueue(result.value);
+      },
+      async cancel(reason) {
+        // Propagate cancellation to readable and iterator
+        if (typeof iterator.throw == "function") {
+          try {
+            await iterator.throw(reason);
+          } catch {
+            /* `iterator.throw()` always throws on site. We catch it. */
+          }
+        }
+        await readable.cancel(reason);
+      },
+    }, readableStrategy),
```

Comment:
This part looks similar to `readableStreamFromIterable` except some error handlings. How about updating `readableStreamFromIterable` to this implementation and using it here?

Reply:
I think it is possible to take the common parts of the two functions and integrate them into a another internal function.
Is this what you meant?

```ts
export function readableStreamFromIterable<T>(
  iterable: Iterable<T> | AsyncIterable<T>,
): ReadableStream<T> {
  return internalReadableStreamFromIterable(iterable);
}

export function transformStreamFromGenerator<I, O>(
  transformer: (src: ReadableStream<I>) => Iterable<O> | AsyncIterable<O>,
  writableStrategy?: QueuingStrategy<I>,
  readableStrategy?: QueuingStrategy<O>,
): TransformStream<I, O> {
  const {
    writable,
    readable,
  } = new TransformStream<I, I>(undefined, writableStrategy);

  const iterable = transformer(readable);
  return {
    writable,
    readable: internalReadableStreamFromIterable(iterable, readableStrategy, readable),
  };
}

function internalReadableStreamFromIterable<T>(
  iterable: Iterable<T> | AsyncIterable<T>,
  readableStrategy?: QueuingStrategy<T>,
  upstream?: ReadableStream,
): ReadableStream<T> {
  // implementation
}
```

Reply:
I imagined like that, but now I see it's a little confusing (I missed `readable` dependency in `internalReadableStreamFromIterable` 😓  )

The current implementation looks just fine to me.